src/HCSR04.cpp

```diff
@@ -7,9 +7,10 @@
 #include "HCSR04.h"
 
 UltraSonicDistanceSensor::UltraSonicDistanceSensor(
-        int triggerPin, int echoPin) {
+        int triggerPin, int echoPin, int maxDistanceCm) {
     this->triggerPin = triggerPin;
     this->echoPin = echoPin;
+    this->maxDistanceCm = maxDistanceCm;
```

Comment:
I would not allow this for now, it will just complicate things. If people ask for it we can add it, but I doubt it will be important.
We could, however, create a constant maxDistanceCm, which can't be set from outside, and is always set to 400cm.

Reply:
Not really more complicated as the default value is set in constructor, both your examples are still running as-is without any modification.

Reply:
True, but I was actually thinking about the case where users provide value bigger than 400cm. What then? We will need to explain that in that case, measurement will be incorrect. On the other hand, why would they provide value lesser than 400cm? To gain some performance? Ok, but is that really useful? I am not sure, that is why I thought, let's see, if people need it, we can add it.

Reply:
Not really more complicated as default value is set in constructor, both of examples compile as-is and I think it's more versatile to use a variable as you can use two or more objects in same code for different sensors with different characteristics.
A global constant prevent usage of different sensors.

Reply:
In case the sensor didn't respond your program is locked for 1 sec, if the distance needs to be checked every second and sensor is defective or distance can't be measured your code will be erratic.

To prevent locking code it must use a timeout in acceptable range.

Reply:
That does make sense! Well, maybe having this option with default set to 400cm, as you suggested, is ok. Being able to have some control over how long will the measuring action maximally take does sound important. But thinking further about it, maybe people would rather provide the timeout directly then, if it is important for them to ensure measuring operation does not take longer than specific time?
I can imagine couple use cases:
1. I know I don't care about distances larger than N cm. In that case, it makes sense to set max distance to N, so each measurement takes less time, and I can do more of them in a period of time, so I get higher resolution.
2. I know that sensor does not support distance larger than N cm.
3. I want to ensure that measurement does not take longer than T time, because I have some requirements in my real time system.

Use cases 1 and 2 are well supported by the change you suggested. There is a tricky edge case -> what if thy specify max distance larger than what their sensor supports? They might be getting weird results. But I guess this is on them. We can mention in documentation that default value provided is for the hc-sr04 sensor and if they are providing different value, they should ensure it does not go beyond maximum of what their sensor can do.

Use case 3 -> well, not really, because they would have to calculate the distance to achieve that time, and the calculation we do to get time based on distance is an "implementation detail" -> we don't want them to bother with that. So if we want to support this use case, we should provide them with a direct way to specify max time. Maybe we could offer them an additional setting, maxMeasurementDurationInMicroSec? Not sure if it should be constructor argument or argument of measure() function. And then we would take use min(maxMeasurementDurationInMicroSec, durationMicroSecForMaxDistance) and use that as a timeout.

Well this is now getting complicated, which is why I initially didn't want to get into it, but your reasons sound good, and if we are doing it let's go all the way! What do you think?

Reply:
For use cases 1 & 2 weird results are impossible, if maxDistance is greater than sensor abilities, sensor will not echo before pulseIn timeout so function will returns -1.

Agreed for use case 3, let me dig into that to find the proper way to do it.

Reply:
I don't think usage of min() is a good idea, if one decide to set a fixed value for timeout we shouldn't alter this value.

I guess using a parameter of the measure method will be hard to maintain if there are several call to this method on the code.

Perhaps we can add a method to specify an absolute timeout, default property is 0 which let automatic value within measure otherwise the fixed value is used.

Do you agree ?

Reply:
@furynick, so sorry for dropping this! I am back if you want to finalize it. Let's see -> so you are proposing we allow them to set a timeout if they want, and in that case, that timeout is respected, not the maxDistance. And it should be an instance method (maybe contrcutors parameter?) -> ok that sounds good.

I am still in favor of using min() though. If we have maxDistance and maxTimeout, then it is clear that they are both max -> it means it might be less! But if I set maxDistance and then I set timeout and suddenly that distance is not respected any more -> well that sounds like something that is unexpected.
What do you say? So I am voting to add maxTimeout next to maxDistance, they are both equivalent in the sense that min() on both of them is used to figure out the actual timeout. And of course, maxTimeout would by default have negative value which we then handle in a special way.

Reply:
New commit pushed, for adding min between maxdistance & timeout, it might be a good idea.
Timeout is set to 0 by default (disabled).